server/src/main/java/org/elasticsearch/health/HealthService.java

```diff
@@ -80,30 +98,96 @@ public List<HealthIndicatorResult> getHealth(@Nullable String indicatorName, boo
         Stream<HealthIndicatorService> filteredIndicators = healthIndicatorServices.stream()
             .filter(service -> indicatorName == null || service.name().equals(indicatorName));
 
-        Stream<HealthIndicatorResult> filteredIndicatorResults;
         if (clusterHealthIsObtainable) {
-            // Calculate remaining indicators
-            filteredIndicatorResults = filteredIndicators.map(service -> service.calculate(explain));
+            if (HealthNode.isEnabled()) {
+                try {
+                    client.execute(FetchHealthInfoCacheAction.INSTANCE, new FetchHealthInfoCacheAction.Request(), new ActionListener<>() {
+                        @Override
+                        public void onResponse(FetchHealthInfoCacheAction.Response response) {
+                            HealthInfo healthInfo = response.getHealthInfo();
+                            combineResultsAndNotifyListener(
+                                listener,
+                                indicatorName,
+                                preflightResults,
+                                filteredIndicators.map(service -> service.calculate(explain, healthInfo)).toList()
+                            );
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            getHealthNoHealthInfo(listener, indicatorName, preflightResults, filteredIndicators.toList(), explain);
+                        }
+                    });
+                } catch (NodeNotConnectedException | HealthNodeNotDiscoveredException e) {
+                    logger.info("Could not fetch data from health node", e);
```

Comment:
This is needed (typically) when the master-is-stable is reporting that there is a stable master node but the master itself is null (since we say the master is stable within the last 30 seconds). I can bump down the level.